### ISSUE TITLE
Shiboken2 import error

### ISSUE THREAD

**Issue by jupiterbjy** [marcelotduarte/cx_Freeze#638]
Using Python 3.8.2, cx-freeze 6.2 ([485eb03](https://github.com/anthony-tuininga/cx_Freeze/commit/485eb0330de934a6418afa42b2eb82f1948cf2d7)).
Same result in 6.1 release.

I'm trying to move from pyinstaller(4.0) but cx_freeze-d executable doesn't run at all with shiboken2 with DLL import error: couldn't find specified module. Got no working solution for this issue with googling, I'm asking it here.

Setup scripts I've used: [1](https://pastebin.com/nC8RrDXm), [2](https://pastebin.com/0m0tfhY9)

![image](https://user-images.githubusercontent.com/26041217/79761442-6c92a080-835c-11ea-80ae-4a4a9668d8b5.png)

cx_freeze still fails to load shiboken2 module even after copy-pasting entire shiboken2 package to lib folder. Any advice? Will provide my codes if needed.

**Comment by marcelotduarte** [comment]
Can you make a sample that I can test?
Maybe a modified version of:
https://github.com/anthony-tuininga/cx_Freeze/tree/master/cx_Freeze/samples/PyQt5

**Comment by jupiterbjy** [comment]
![Screenshot_20200421-181845_VNC Viewer](https://user-images.githubusercontent.com/26041217/79848991-d498c380-83fc-11ea-981f-89744f58b0d2.jpg)

Same result with same sample code - except changing PyQt5 to Pyside2. Sorry for phone capture, as I have no access to PC nor had time to cut this time. Army issues..

**Comment by marcelotduarte** [comment]
I need to reproduce the error, not to view your error. I need a sample.

**Comment by fred816** [comment]
The same thing happened on a computer without Python installed.

**Comment by jupiterbjy** [comment]
> I need to reproduce the error, not to view your error. I need a sample.

>Same result with same sample code - except changing PyQt5 to Pyside2. 

As I said, that is from exact same sample code you've linked. Only changed pyqt5 to pyside2 at import line. Setup.py is untouched.

If you still need it, here it is.

````
import sys
from PySide2.QtWidgets import QApplication, QWidget

app = QApplication(sys.argv)
window = QWidget()
window.setWindowTitle("Simple")
window.show()
app.exec_()
````

**Comment by konserw** [comment]
Hi!
It's probably this: https://bugreports.qt.io/browse/PYSIDE-1257
PySide2 5.14.2 introduced this new method:
```
def _additional_dll_directories(package_dir):
    # Find shiboken2 relative to the package directory.
    root = os.path.dirname(package_dir)
    shiboken2 = os.path.join(root, 'shiboken2')
    if os.path.isdir(shiboken2): # Standard case, only shiboken2 is needed
        return [shiboken2]
    # The below code is for the build process when generate_pyi.py
    # is executed in the build directory. We need libpyside and Qt in addition.
    shiboken2 = os.path.join(os.path.dirname(root), 'shiboken2', 'libshiboken')
    if not os.path.isdir(shiboken2):
        raise ImportError(shiboken2 + ' does not exist')
    result = [shiboken2, os.path.join(root, 'libpyside')]
    for path in os.environ.get('PATH').split(';'):
        if path:
             if os.path.exists(os.path.join(path, 'qmake.exe')):
                 result.append(path)
                 break
    return result
```
Which breaks cx_forzen apps (at least on windows). Basically in normal flow (not frozen) ` if os.path.isdir(shiboken2):` evaluates to True, so method returns in next line and all is good; but if in frozen case it evaluates to False (as shiboken dir is not where it looks for it), so it looks for linux lib or something, so it rises exception as tha can't be found either. 

It's regression in PySide code, but I'm not sure if it have to be fixed there, or can be worked around here. Please cooperate with them on this :) 

I hope this is helpful.

**Comment by konserw** [comment]
And @jupiterbjy for now just switch back to PySide2 5.14.1

**Comment by jupiterbjy** [comment]
That's crystal clear, thanks!

Even though Pyinstaller worked around for this, making exception for one specific module doesn't sound legit. Just seeing how slow Pyinstaller is a good reason not to do so.

On that regard - Should I close issue or leave this open?

**Comment by marcelotduarte** [comment]
I'll take a look. Leave this open.

**Comment by marcelotduarte** [comment]
Look what I found out.
On Linux, with PySide2 5.14.1 build, upgrade, and rebuild, it gives an error, but if you delete the 'build' and rebuild, it works.
On Windows, something like that, but with a clean build, the python3.dll file is missing which must be placed together with the executable. I will make a patch for this.
On Windows, I tested with Py 3.7 32bits.
Can you test if copying the python3.dll?

**Comment by jupiterbjy** [comment]
If I recall correctly there was python38.dll with exe.  Forgot to mention this - I'm using 64bit python 3.8.2.

Since I have midnight drill today, I'll try freezing with 32bit python tomorrow. I'm terribly sorry for inconvienence. No access to PC for 27 hours from now on.

EDIT: found [Screenshot](https://imgur.com/a/zBCQNPX) of exe freshly build before with python38.dll alongside with executable.

**Comment by marcelotduarte** [comment]
No inconvenience. You can test with python 32, 64 bit. The results should works.
I already had a patch. Later I'll send a PR.
~~You should copy the python3.dll from the installation of your python to the build directory, along with python38.dll.~~ (The PR solves this).

**Comment by marcelotduarte** [comment]
Can you test the PR? (only if you have a C compiler)

`pip install -U git+https://github.com/anthony-tuininga/cx_Freeze.git@refs/pull/640/head`

**Comment by jupiterbjy** [comment]
Reinstalled C++ 14 Build tool due to basetsd.h issue - And now cx-freeze works like charm. Thanks! 

And to add, I didn't know python3.dll differs with python38.dll, should've tried it earlier..

I guess this issue closes automatically upon merging?

**Comment by marcelotduarte** [comment]
> I guess this issue closes automatically upon merging?

yes

**Comment by marcelotduarte** [comment]
@konserw Maybe the change you reported in pyside2 broke the code, but my PR solve this issue, then I guess that you can inform the pyside2 team.
Can you test the PR? (only if you have a C compiler)

`pip install -U git+https://github.com/anthony-tuininga/cx_Freeze.git@refs/pull/640/head`

**Comment by konserw** [comment]
Hi,
Just like I've suspected adding python3.dll doesn't solve missing shiboken.dll issue. See: 
![obraz](https://user-images.githubusercontent.com/3592272/80789452-540c5c80-8b8c-11ea-87dd-da8890a43239.png)

@jupiterbjy shiboken issue is gone for you? On which pyside version?

**Comment by marcelotduarte** [comment]
@konserw In my tests, with PySide2 5.14.1 build, upgrade to 5.14.2, and rebuild, it gives an error, but if you delete the build directory and rebuild, it works.

**Comment by jupiterbjy** [comment]
Same Configuration, built cx-freeze using pull request 640. Freezed inside venv.

By the way, there was newly released PySide2 5.14.2.1, will try freezing that tonight.

EDIT: PySide2 5.14.2.1 also works well with cx_freeze.
